src/ipc/ipc4/handler.c

```diff
@@ -337,28 +314,119 @@ int set_pipeline_state(struct ipc_comp_dev *ppl_icd, uint32_t cmd,
 	case SOF_IPC4_PIPELINE_STATE_SAVED:
 	case SOF_IPC4_PIPELINE_STATE_ERROR_STOP:
 	default:
-		ipc_cmd_err(&ipc_tr, "ipc: unsupported trigger cmd 0x%x", cmd);
+		ipc_cmd_err(&ipc_tr, "pipeline %d: unsupported trigger cmd: %d",
+			    ppl_icd->id, cmd);
 		return IPC4_INVALID_REQUEST;
 	}
 
-	tr_dbg(&ipc_tr, "ipc4: set pipeline %d new cmd state %x: from state %x", ppl_icd->id,
-	       cmd, status);
+	return ret;
+}
+
+static int ipc4_pipeline_trigger(struct ipc_comp_dev *ppl_icd, uint32_t cmd,
+				 bool *delayed)
+{
+	struct ipc_comp_dev *host;
+	int status;
+	int ret;
+
+	status = ppl_icd->pipeline->status;
+	tr_dbg(&ipc_tr, "pipeline %d: initial state: %d, cmd: %d", ppl_icd->id,
+	       status, cmd);
+
+	if (status == COMP_STATE_INIT)
+		return 0;
+
+	host = pipeline_get_host_dev(ppl_icd);
+	if (!host)
+		return IPC4_INVALID_RESOURCE_ID;
+
+	switch (cmd) {
+	case SOF_IPC4_PIPELINE_STATE_RUNNING:
+		/* init params when pipeline is complete or reset */
+		switch (status) {
+		case COMP_STATE_ACTIVE:
+			/* nothing to do if the pipeline is already running */
+			return 0;
+		case COMP_STATE_READY:
+		case COMP_STATE_PREPARE:
+			cmd = COMP_TRIGGER_PRE_START;
```

Comment:
it seems to me this does way too much. Can we just pass `cmd` by reference to `prepare()` and **keep** the current top part of `set_pipeline_state()` 1-to-1 with only replacing `cmd` with `*cmd` and then use that `cmd` for the trigger part. I think that way the diff should become *much* smaller and a lot clearer and the chance of an accidental change will be much lower.